### PR TITLE
Add userguide for AdaSum

### PR DESCRIPTION
**Pull request by ashbhandare**
This PR adds a userguide for using Adasum.
This userguide introduces the algorithm and has instructions on using Adasum with different hardware
and configurations.
This also includes a small example optimization problem to better understand the difference between Averaging and Adasum, helping in understanding the benefits of the algorithm.

### CODE REVIEW

**Comment by tgaddair** [comment]
@ashbhandare can you rebase to fix unit tests?

**Comment by sblotner** [comment]
Thank you for adding this guide with examples, @ashbhandare. Are you looking to publish it to the Horovod doc site https://horovod.readthedocs.io/en/latest/? 

If so, please note that the documentation is generated from .rst files, like this: https://raw.githubusercontent.com/horovod/horovod/master/docs/mxnet.rst

If you'd like to publish the content to the Horovod doc site, please:

1. Convert  the user guide file to .rst (here is 
2. Reference the user guide file from table of contents (toctree) in index.rst: https://raw.githubusercontent.com/horovod/horovod/master/docs/index.rst

**Comment by ashbhandare** [comment]
Thank you, @tgaddair I have rebased.
@sblotner  I have changed the format to rst and added the reference to index.rst

File: docs/adasum_user_guide.rst [hunk 1]
```diff
@@ -0,0 +1,214 @@
+.. inclusion-marker-start-do-not-remove
+
+AdaSum with Horovod
+=================
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Please extend underline (==) to cover the end of "Horovod"

File: docs/adasum_user_guide.rst [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+The Adaptive Summation, or AdaSum is a novel algorithm for improving distributed data parallel training of Deep Learning models. This improvement can be seen in different ways- reducing the number steps to achieve the same accuracy in some cases and allowing to scale to more training workers without penalizing learning rate and convergence stability.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: - Add a comma after "or AdaSum"
- Replace "ways-" with "ways: "
- "allowing to --> allowing you to"

File: docs/adasum_user_guide.rst [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+.. Contents::
+
+|
+
+
+Introduction to the AdaSum Algorithm
+======================================
+
+
+Scaling DNN training to many GPUs always comes at a convergence degradation. This is because with larger batch sizes, gradients are averaged and the learning rate per example is smaller. To address this, learning rate is usually scaled up but this can lead to divergence of model parameters. AdaSum addresses these two issues without introducing any hyper-parameter.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: - Add a comma after "scaled up"
- "hyper-parameter" --> "hyperparameter"

File: docs/adasum_user_guide.rst [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+Suppose there are two almost-parallel gradients from two different GPUs, g1 and g2, and they need to be reduced as shown in the figure below. The two common practices for reductions are g1+g2, the gray vector or (g1+g2)/2, the green vector. g1+g2 may cause divergence of the model since it is effectively moving in the direction of g1 or g2 by two times the magnitude of g1 or g2. Therefore, generally (g1+g2)/2 is safer and more desired. Note that (g1+g2)/2 penalizes both the components g1 and g2 equally.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Add a comma after "gray vector"

File: docs/adasum_user_guide.rst [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+This formula reduces to a sum when g1 and g2 are orthogonal and an average when g1 and g2 are parallel.
+
+This idea extends to many gradients as well. Suppose there are 2\^n gradients coming from 2\^n different GPUs. AdaSum inductively takes pairs of gradients and reduces them using the method above until all of them are reduced into one gradient. Thus, AdaSum needs the number of nodes to be a power of 2 in the current implementation.
+
+
+The Distributed Optimizer for AdaSum
+======================================
+
+
+AdaSum uses the Distributed AdaSum Optimizer to update the weights of the model after each step. In the usual data-parallel training scenario, the gradients are calculated independently by backpropagating on all the nodes, doing a reduce(averaging the gradients) so that all the nodes now have the same gradients, and then updating the weights of the model.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Add a space after "reduce"

File: docs/adasum_user_guide.rst [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+The distributed optimizer for AdaSum first obtains the local gradients from the backpropagation step from the current local mini batch. Instead of performing the reduce at this point, it applies the optimization function to the local gradients to perform the weight update. Then, the delta, which is the differencein the weights before and after the update is obtained, which is then reduced instead of the gradients. Once all the workers have the same delta, the weight update step is then performed as the sum of the initial weights and delta.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: - Add a space after "difference"

File: docs/adasum_user_guide.rst [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+Since the nature of AdaSum requires it to operate on the full magnitude of the gradient, the newly added distributed optimizer uses the difference in magnitude of weights between before and after the optimizer performs a step to deliver a more accurate estimation.
+
+
+Installation and Usage Instructions
+=====================================
+
+
+AdaSum can be used and experimented with Horovod and Pytorch/TensorFlow.
+
+In addition, there are two options of using AdaSum with Horovod- with Message Passing Interface(MPI) and with `NCCL <https://developer.nvidia.com/nccl>`_.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: - "Horovod-" --> "Horovod: "
- Add a space after "Interface"

File: docs/adasum_user_guide.rst [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+
+Installation and Usage Instructions
+=====================================
+
+
+AdaSum can be used and experimented with Horovod and Pytorch/TensorFlow.
+
+In addition, there are two options of using AdaSum with Horovod- with Message Passing Interface(MPI) and with `NCCL <https://developer.nvidia.com/nccl>`_. 
+It can be noted that any valid implementation of MPI can be used, but AdaSum has been tested with `OpenMPI <https://www.open-mpi.org/>`_ and `IntelMPI <https://software.intel.com/en-us/mpi-library>`_.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Remove "It can be noted that"

File: docs/adasum_user_guide.rst [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+-   Pytorch-1.2.0 OR
+
+-   Tensorflow-1.14.0
+
+-   Horovod-0.18.2
+
+*Using NCCL:*
+
+If **HOROVOD_GPU_ALLREDUCE=NCCL** flag is used to compile Horovod, NCCL is used instead. In this case, NCCL will be used for intra-node communication, and AdaSum will be used for inter-node communication.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: - If --> If the

File: docs/adasum_user_guide.rst [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+Ring mode is currently supported only on **DGX1** nodes having 8 GPUs each.
+
+.. image:: media/4920a765a77fa6eeca28c5aceaa405ec.png
+
+Hierarchical
+--------------------------
+
+In cases where the hardware does not support Ring mode, but throughput higher than that of the pure CPU mode is desired, the hierarchical mode can be used instead.
+
+The hierarchical mode functions similar to the Ring mode, except for using NCCL to do regular averaging intra-node, instead of using CUDA-aware MPI to do AdaSum like ring. Note that hierarchical also works on any hardware configuration, and is not limited to DGX1s.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: do AdaSum like ---> do an AdaSum-like

File: docs/adasum_user_guide.rst [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+A new distributed optimizer has been added to both TensorFlow and Pytorch to support the AdaSum algorithm.
+
+An optional parameter op has been added to DistributedOptimizer and allreduce API for users to specify which operation to perform.
+When **op=hvd.AdaSum** is specified, the new optimizer will be used.
+
+TensorFlow
+--------------------------
+
+-   DistributedOptimizer
+.. code-block:: python
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Add a blank line before this line

File: docs/adasum_user_guide.rst [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+ opt = tf.train.AdamOptimizer(0.001)
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Indent the contents of the code block 4 spaces

File: docs/adasum_user_guide.rst [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+ opt = hvd.DistributedOptimizer(opt, backward_passes_per_step=5, op=hvd.AdaSum)
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Indent the contents of the code block 4 spaces

File: docs/adasum_user_guide.rst [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+-   Allreduce
+.. code-block:: python
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: add a blank line before this line

File: docs/adasum_user_guide.rst [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    
+ hvd.allreduce(tensor, op=hvd.AdaSum)
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Indent the contents of the code block 4 spaces

File: docs/adasum_user_guide.rst [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+Pytorch
+--------------------------
+
+-   DistributedOptimizer
+
+.. code-block:: python
+
+ optimizer = optim.SGD(model.parameters(), lr=args.lr, momentum=args.momentum)
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Indent the contents of the code block 4 spaces

File: docs/adasum_user_guide.rst [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+ optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters(), compression=compression, backward_passes_per_step = 5, op=hvd.AdaSum)
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Indent the contents of the code block 4 spaces

File: docs/adasum_user_guide.rst [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+
+-   Allreduce
+
+.. code-block:: python
+
+ hvd.allreduce(tensor, op=hvd.AdaSum)
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: Indent the contents of the code block 4 spaces

File: docs/adasum_user_guide.rst [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+
+In order to estimate the coefficients, Stochastic Gradient Descent is used. The training is stopped once the gradients are zero for two consecutive runs. This optimization can be run over a range of learning rates, number of workers and data range (set by x_max). This can also be modified to a cubic optimization problem.
+
+This experiment can be run through the jupyter notebook `adasum_bench.ipynb <adasum_bench.ipynb>`_, with the models being defined in `adasum_small_model.py <adasum_small_model.py>`_.
+
+On running experiments with different number of workers, we can draw the following conclusions for this simple scenario with plain SGD as the optimizer:
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: with different --> with a different

File: docs/adasum_user_guide.rst [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+Key Takeaways
+===============
+
+|
+
+-   AdaSum ensures correct convergence behavior even with large effective batch sizes.
+
+-   As the number of ranks scales up, the learning rate does not need to be scaled linearly if using CPU to do AdaSum reduction. A good scaling factor would be between 2\-2.5 over the best learning rate for a single worker.
+
+-   If HOROVOD_GPU_ALLREDUCE=NCCL flag is used to compile Horovod, the learning rate that should be used is equal to the best learning rate for a single	worker (GPU) scaled by the number of GPUs locally on a node. On very large	clusters, scaling this even more by another factor of 1.5\-2.0x might give	better results but is not guaranteed and should be tried only if scaling by just the local size is not sufficient for good convergence.
```
Comment (sblotner) [inline docs/adasum_user_guide.rst]: If --> If the

**Comment by ashbhandare** [comment]
> Thanks for updating the file format! Added some editorial comments

Thank you for the comments! Have updated.

**Comment by sblotner** [comment]
Thanks! Looking at the build log, there may be an issue with including a notebook under /docs - might need to move it under /examples. 

Specifically, from Build Docs: https://buildkite.com/horovod/horovod/builds/2352#5d74e878-5e85-49c0-b8b2-f244ff251993

`
Notebook error:
--
  | ImportError in adasum_bench.ipynb:
  | `nbconvert --execute` requires the jupyter_client package: `pip install jupyter_client`

`

**Comment by ashbhandare** [comment]
> Thanks! Looking at the build log, there may be an issue with including a notebook under /docs - might need to move it under /examples.
> 
> Specifically, from Build Docs: https://buildkite.com/horovod/horovod/builds/2352#5d74e878-5e85-49c0-b8b2-f244ff251993
> 
> ## `
> Notebook error:
>   | ImportError in adasum_bench.ipynb:
>   | `nbconvert --execute` requires the jupyter_client package: `pip install jupyter_client`
> 
> `

Thanks! Have moved the files to the examples folder.

File: docs/adasum_user_guide.rst [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+
+TensorFlow
+--------------------------
+
+-   DistributedOptimizer
+
+.. code-block:: python
+
+    opt = tf.train.AdamOptimizer(0.001)
+    opt = hvd.DistributedOptimizer(opt, backward_passes_per_step=5, op=hvd.AdaSum)
```
Comment (tgaddair) [inline docs/adasum_user_guide.rst]: Can you comment on when it makes sense to set `backward_passes_per_step`?

File: docs/adasum_user_guide.rst [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+
+AdaSum can be used and experimented with Horovod and Pytorch/TensorFlow.
+
+In addition, there are two options of using AdaSum with Horovod: with Message Passing Interface (MPI) and with `NCCL <https://developer.nvidia.com/nccl>`_. 
+Any valid implementation of MPI can be used, but AdaSum has been tested with `OpenMPI <https://www.open-mpi.org/>`_ and `IntelMPI <https://software.intel.com/en-us/mpi-library>`_.
+
+Setting up the environment
+--------------------------
+
+Below are the requirements for running Horovod with AdaSum:
```
[the remark this answers is not in the document]
Comment (ashbhandare) [reply docs/adasum_user_guide.rst:64]: I have added min requirements

File: docs/adasum_user_guide.rst (hunk 21 above)
Comment (ashbhandare) [reply docs/adasum_user_guide.rst:138]: AdaSum is useful for scaling to large batch sizes, and that can be done with a limited GPU memory by using gradient accumulation using `backward_passes_per_step`. I have added the same explanation in the userguide, thanks!

File: docs/adasum_user_guide.rst [hunk 23]
```diff
@@ -63,19 +63,17 @@ Setting up the environment
...
+-   OpenMPI >= 3.0
 
--   NCCL – 2.4
+-   NCCL >= 2.0
 
--   Python 3.7.5
+-   Pytorch >= 1.2.0 OR
 
--   Pytorch-1.2.0 OR
+-   Tensorflow >= 1.14.0
```
Comment (tgaddair) [inline docs/adasum_user_guide.rst:74]: @ashbhandare, does Adasum support TensorFlow 2?  Last I checked it was only for TensorFlow 1.

Comment (ashbhandare) [reply docs/adasum_user_guide.rst]: That is correct, added the requirement.

**Comment by ashbhandare** [comment]
> LGTM! Will land once the tests pass.

Great! Looks like the tests have passed.

**Comment by liuyunfeng2016** [comment]
@ashbhandare  @tgaddair  @sblotner  hello,I found that hvd.AdaSum is incorrect in Horovod 0.21.0. It should be hvd.Adasum.
